LDMP/models/algorithms_delegate.py

```diff
@@ -0,0 +1,263 @@
+"""
+
+__author__ = 'Luigi Pirelli / Kartoza'
+__date__ = '2021-03-03'
+
+from functools import partial
+from typing import Optional, Union
+from qgis.PyQt.QtCore import (
+    QModelIndex,
+    Qt,
+    QCoreApplication,
+    QObject,
+    pyqtSignal,
+    QRectF,
+    QAbstractItemModel
+)
+from qgis.PyQt.QtWidgets import (
+    QStyleOptionViewItem,
+    QToolButton,
+    QMenu,
+    QStyledItemDelegate,
+    QItemDelegate,
+    QWidget,
+    QAction
+)
+from qgis.PyQt.QtGui import (
+    QPainter,
+    QPixmap
+)
+from qgis.PyQt.QtSvg import (
+    QSvgRenderer
+)
+
+from LDMP.models.algorithms import (
+    AlgorithmGroup,
+    AlgorithmDescriptor,
+    AlgorithmNodeType,
+    AlgorithmRunMode
+)
+from LDMP import __version__, log, tr
+
+
+# Icons used for animation:
+# inspired by https://wiki.python.org/moin/PyQt/Animated%20items%20using%20delegates
+_icon0_xpm = [
+    "16 16 4 1",
+    ".  c None",
+    "s  c #cccc00",
+    "x  c #ffffff",
+    "   c #000000",
+    "................",
+    ".              .",
+    ". xxxxxxxxxxxx .",
+    ". x          x .",
+    ".. x ssssss x ..",
+    "... x ssss x ...",
+    ".... x ss x ....",
+    "..... x  x .....",
+    "..... x  x .....",
+    ".... x    x ....",
+    "... x      x ...",
+    ".. x        x ..",
+    ". x          x .",
+    ". xxxxxxxxxxxx .",
+    ".              .",
+    "................"]
+
+_icon1_xpm = [
+    "16 16 4 1",
+    ".  c None",
+    "s  c #cccc00",
+    "x  c #ffffff",
+    "   c #000000",
+    "................",
+    ".              .",
+    ". xxxxxxxxxxxx .",
+    ". x          x .",
+    ".. x s    s x ..",
+    "... x ssss x ...",
+    ".... x ss x ....",
+    "..... x  x .....",
+    "..... x  x .....",
+    ".... x ss x ....",
+    "... x  ss  x ...",
+    ".. x        x ..",
+    ". x          x .",
+    ". xxxxxxxxxxxx .",
+    ".              .",
+    "................"]
+
+_icon2_xpm = [
+    "16 16 4 1",
+    ".  c None",
+    "s  c #cccc00",
+    "x  c #ffffff",
+    "   c #000000",
+    "................",
+    ".              .",
+    ". xxxxxxxxxxxx .",
+    ". x          x .",
+    ".. x        x ..",
+    "... x ssss x ...",
+    ".... x ss x ....",
+    "..... x  x .....",
+    "..... x  x .....",
+    ".... x ss x ....",
+    "... x  ss  x ...",
+    ".. x   ss   x ..",
+    ". x          x .",
+    ". xxxxxxxxxxxx .",
+    ".              .",
+    "................"]
+
+_icon3_xpm = [
+    "16 16 4 1",
+    ".  c None",
+    "s  c #cccc00",
+    "x  c #ffffff",
+    "   c #000000",
+    "................",
+    ".              .",
+    ". xxxxxxxxxxxx .",
+    ". x          x .",
+    ".. x        x ..",
+    "... x      x ...",
+    ".... x    x ....",
+    "..... x  x .....",
+    "..... x  x .....",
+    ".... x ss x ....",
+    "... x ssss x ...",
+    ".. x ssssss x ..",
+    ". x          x .",
+    ". xxxxxxxxxxxx .",
+    ".              .",
+    "................"]
+
+class AlgorithmItemDelegate(QStyledItemDelegate):
```

Comment:
I'm not sure using a delegate is the best option here. Perhaps we can use [QAbstractItemView.setIndexWidget](https://doc.qt.io/qt-5/qabstractitemview.html#setIndexWidget) instead and just use a custom widget?

My main reasoning for this is that I don't like the current behavior where we need to click once on the tree row just to activate it and make the delegate be rendered in editing mode before we can then click the button for running an algorithm, which seems clumsy from a UX point of view.

Maybe there is another way to do it with a delegate, I'm not sure. In any case, what I want to prevent is the user not knowing how many times he needs to click on the button to get it to run

Reply:
I tried to avoid setIndexWidget because it imply low performances and should be avoided... it's not out case with few element of the tree, BTW I'll try to use setItemWidget.
I agree that the user shouldn't have to click to activate editing